p2p/src/net/mod.rs

```diff
@@ -196,43 +35,48 @@ pub trait NetworkingService {
     ///
     /// For an implementation built on libp2p, the address format is:
     ///     `/ip4/0.0.0.0/tcp/8888/p2p/<peer ID>`
-    type Address: Send + Sync + Debug + PartialEq + Eq + Hash + Clone + ToString;
+    type Address: Clone + Debug + Eq + Hash + PartialEq + Send + Sync + ToString;
 
     /// Unique ID assigned to a peer on the network
-    type PeerId: Send + Copy + PartialEq + Eq + Hash + Debug + Sync + ToString;
+    type PeerId: Copy + Debug + Eq + Hash + PartialEq + Send + Sync + ToString;
 
-    // TODO:
-    type RequestId: Send + Debug + Eq + Hash + Sync;
+    /// Unique ID assigned to each received request
+    type RequestId: Debug + Eq + Hash + Send + Sync;
 
     /// Enum of different peer discovery strategies that the implementation provides
     type DiscoveryStrategy;
 
     /// Id that identifies a protocol
-    type ProtocolId: Debug + Send + Clone + PartialEq;
+    type ProtocolId: Clone + Debug + PartialEq + Send;
```

Comment:
There's absolutely no reason I'm aware of to require PartialEq and not Eq unless we're dealing with floating point numbers. Either both or nothing, please.

Reply:
Btw, this problem exists in many places, not just here.

Reply:
I'm not sure what you mean. I removed `PartialEq` from `PeerId` and `Address` as it was not used but `ProtocolId` needs it. It doesn't need `Eq` though so I'm not entirely sure what you're saying.

Reply:
As far as I can tell, this is only used in ad-hoc `PartialEq` instance for `PeerInfo` in libp2p tests.

A couple of observations:
1. The `PeerInfo` impl could be restricted locally using a `where` clause, instead of restricting `ProtocolId` globally
2. It would be cleaner for `PeerInfo` to provide `Eq` as well provided corresponding `NetworkingService::ProtocolId` implements `Eq`, but I am not too bothered by that since it is only in tests
3. The impl is required for `assert_eq!`. While this strictly only requires `PartialEq`, it is reasonable to expect total equality here. I don't see a simple way of adding the additional restriction and also not probably not wort hit.
4. Trait impls are visible "globally", you just need to import the trait and have a value of the corresponding type. This applies even if the trait impl itself is buried in a submodule. So it's better to get trait impls right.
5. However, the `PeerInfo` impl is in test module and as such is only compiled in when testing, so (4) only becomes an issue when someone moves the impl out of the test module without realizing full implications
6. Having laid out all the possible issues here, I still think this is a very low-risk issue.

Reply:
IIRC the reason why `PeerInfo` manually implemented `PartialEq` was because it contained `Arc<T>`. I'm right now in the process of trying to remove that because hopefully it's not needed. I agree that `PartialEq` implementation should be visible globally and I will make it visible, if the `ProtocolId`/`PeerInfo` in its current form survive the maintenance period. We can revisit this issue when the `PeerManager` implementation is in better condition and it's more clear what is needed.